src/odemis/gui/comp/viewport.py

```diff
@@ -1065,7 +1066,334 @@ def _on_peak_method(self, state):
         pass
 
 
-class PointSpectrumViewport(PlotViewport):
+class NavigablePlotViewport(PlotViewport):
+    """ Abstract Class for displaying plotted data
+
+    Unlike PlotViewport, the NavigablePlotViewport can be manipulated by the user.
+    The plot can be zoomed and panned
+        - Mousewheel over an axis zooms that axis.
+        - Left mouse drag over an axis pans that axis.
+        - Mousewheel over the canvas zooms the horizontal axis
+        - Middle mouse button drag on the canvas pans the plot
+
+    This is controlled by VA's
+    .hrange: Tuple VA with the horizontal scale range
+    .vrange: Tuple VA with the vertical scale range
+    .hrange_lock, .vrange_lock: Bool VA that locks H/V axes from manipulation
+
+    Note: clamping of the ranges is handle by the mouse events, because
+    the clamping behavior must function differently for pans, drags, and scrolling
+
+    """
+    __metaclass__ = no_conflict.classmaker(right_metas=(ABCMeta,))
+
+    # Default class
+    canvas_class = miccanvas.NavigableBarPlotCanvas
+    bottom_legend_class = AxisLegend
+    left_legend_class = AxisLegend
+
+    def __init__(self, *args, **kwargs):
+        super(NavigablePlotViewport, self).__init__(*args, **kwargs)
+
+        self.hrange = model.TupleVA(setter=self.update_hrange)
+        self.bottom_legend.tick_spacing = 80
+        self.hrange_lock = model.BooleanVA(False)
+        self.bottom_legend.lock_va = self.hrange_lock
+
+        self.vrange = model.TupleVA(setter=self.update_vrange)
+        self.vrange_lock = model.BooleanVA(False)
+        self.left_legend.lock_va = self.vrange_lock
+
+        # Indicate a left mouse button drag in the canvas
+        # Note: *only* use it indicate that the *canvas* is performing an operation related to
+        # dragging!
+        self._dragging = False
+        self._hdrag_scale_factor = -0.1
```

Comment:
What are these scale factors about? I somehow imagine that the "-" is to make sure that it goes in the same direction as the mouse, while dragging with the legend causes the opposite direction move... But they all have this one, so I'm confused. And why dividing by 10? Is that to do a small move? If so, why not making sure that it respects the position of the mouse?

Reply:
Hi. I found that somehow the canvas dragged too fast without dividing by 10, and it was a bit awkward for the user to control.